casting to a type with template constructor

The following code fails to compile both on MSVC 2015 and on g++ 4.9.4.
```C++
struct S
{
    template<typename T>
    S(const T& str): m_str(str)
    {}

    std::string m_str;
};

void foo(std::vector<S>&);

namespace py = pybind11;

PYBIND11_PLUGIN(example){
    py::module m("example", "pybind11 example plugin");
    m.def("foo", &foo, "A function which takes vector of S");
    return m.ptr();
}
```

The error on MSVC is
```

1>C:\Users\Dvir\Documents\GitHub\pybind11-test\test.cpp(9): error C2664: 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>::basic_string(const std::basic_string<char,std::char_traits<char>,std::allocator<char>> &)': cannot convert argument 1 from 'const pybind11::detail::type_caster<S,void>' to 'std::initializer_list<_Elem>'
1>          with
1>          [
1>              _Elem=char
1>          ]
1>  C:\Users\Dvir\Documents\GitHub\pybind11-test\test.cpp(9): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
1>  C:\Users\Dvir\Documents\GitHub\pybind11-test\pybind11\include\pybind11/stl.h(124): note: see reference to function template instantiation 'S::S<pybind11::detail::type_caster<S,void>>(const T &)' being compiled
1>          with
1>          [
1>              T=pybind11::detail::type_caster<S,void>
1>          ]
```

If I replace the cast on https://github.com/pybind/pybind11/blob/master/include/pybind11/stl.h#L124 with a cast to `(Value&)` compilation succeeds.
Is there a reason for this cast?

--- a/include/pybind11/stl.h
+++ b/include/pybind11/stl.h
@@ -53,7 +53,7 @@ template <typename Type, typename Key> struct set_caster {
         for (auto entry : s) {
             if (!conv.load(entry, convert))
                 return false;
-            value.insert((Key) conv);
+            value.insert(conv.operator typename key_conv::template cast_op_type<Key>());
         }
         return true;
     }
@@ -87,7 +87,9 @@ template <typename Type, typename Key, typename Value> struct map_caster {
             if (!kconv.load(it.first.ptr(), convert) ||
                 !vconv.load(it.second.ptr(), convert))
                 return false;
-            value.emplace((Key) kconv, (Value) vconv);
+            value.emplace(
+                    kconv.operator typename   key_conv::template cast_op_type<Key>(),
+                    vconv.operator typename value_conv::template cast_op_type<Value>());
         }
         return true;
     }
@@ -121,7 +123,7 @@ template <typename Type, typename Value> struct list_caster {
         for (auto it : s) {
             if (!conv.load(it, convert))
                 return false;
-            value.push_back((Value) conv);
+            value.push_back(conv.operator typename value_conv::template cast_op_type<Value>());
         }
         return true;
     }
@@ -167,7 +169,7 @@ template <typename Type, size_t Size> struct type_caster<std::array<Type, Size>>
         for (auto it : l) {
             if (!conv.load(it, convert))
                 return false;
-            value[ctr++] = (Type) conv;
+            value[ctr++] = conv.operator typename value_conv::template cast_op_type<Type>();
         }
         return true;
     }
@@ -200,13 +202,12 @@ template <typename Key, typename Value, typename Hash, typename Equal, typename 
 
 // This type caster is intended to be used for std::optional and std::experimental::optional
 template<typename T> struct optional_caster {
-    using value_type = typename intrinsic_type<typename T::value_type>::type;
-    using caster_type = type_caster<value_type>;
+    using value_conv = make_caster<typename T::value_type>;
 
     static handle cast(const T& src, return_value_policy policy, handle parent) {
         if (!src)
             return none().inc_ref();
-        return caster_type::cast(*src, policy, parent);
+        return value_conv::cast(*src, policy, parent);
     }
 
     bool load(handle src, bool convert) {
@@ -215,18 +216,16 @@ template<typename T> struct optional_caster {
         } else if (src.is_none()) {
             value = {};  // nullopt
             return true;
-        } else if (!inner.load(src, convert)) {
-            return false;
-        } else {
-            value.emplace(static_cast<const value_type&>(inner));
-            return true;
         }
+        value_conv inner_caster;
+        if (!inner_caster.load(src, convert))
+            return false;
+
+        value.emplace(inner_caster.operator typename value_conv::template cast_op_type<typename T::value_type>());
+        return true;
     }
 
-    PYBIND11_TYPE_CASTER(T, _("Optional[") + caster_type::name() + _("]"));
-
-private:
-    caster_type inner;
+    PYBIND11_TYPE_CASTER(T, _("Optional[") + value_conv::name() + _("]"));
 };
 
 #if PYBIND11_HAS_OPTIONAL

--- a/tests/test_issues.cpp
+++ b/tests/test_issues.cpp
@@ -64,6 +64,15 @@ private:
 };
 PYBIND11_DECLARE_HOLDER_TYPE(T, custom_unique_ptr<T>);
 
+/// Issue #528: templated constructor
+struct TplConstrClass {
+    template <typename T> TplConstrClass(const T &arg) : str{arg} {}
+    std::string str;
+    bool operator==(const TplConstrClass &t) const { return t.str == str; }
+};
+namespace std {
+template <> struct hash<TplConstrClass> { size_t operator()(const TplConstrClass &t) const { return std::hash<std::string>()(t.str); } };
+}
 
 
 void init_issues(py::module &m) {
@@ -371,6 +380,16 @@ void init_issues(py::module &m) {
     py::class_<MyDerived, MyBase>(m2, "MyDerived")
         .def_static("make", &MyDerived::make)
         .def_static("make2", &MyDerived::make);
+
+    /// Issue #528: templated constructor
+    m2.def("tpl_constr_vector", [](std::vector<TplConstrClass> &) {});
+    m2.def("tpl_constr_map", [](std::unordered_map<TplConstrClass, TplConstrClass> &) {});
+    m2.def("tpl_constr_set", [](std::unordered_set<TplConstrClass> &) {});
+#if defined(PYBIND11_HAS_OPTIONAL)
+    m2.def("tpl_constr_optional", [](std::optional<TplConstrClass> &) {});
+#elif defined(PYBIND11_HAS_EXP_OPTIONAL)
+    m2.def("tpl_constr_optional", [](std::experimental::optional<TplConstrClass> &) {});
+#endif
 }
 
 // MSVC workaround: trying to use a lambda here crashes MSCV

